flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/recovery/RescalingStreamTaskNetworkInput.java

```diff
@@ -0,0 +1,280 @@
+
+    private static final Logger LOG =
+            LoggerFactory.getLogger(RescalingStreamTaskNetworkInput.class);
+    private final IOManager ioManager;
+
+    private RescalingStreamTaskNetworkInput(
+            CheckpointedInputGate checkpointedInputGate,
+            TypeSerializer<T> inputSerializer,
+            IOManager ioManager,
+            StatusWatermarkValve statusWatermarkValve,
+            int inputIndex,
+            InflightDataRescalingDescriptor inflightDataRescalingDescriptor,
+            Function<Integer, StreamPartitioner<?>> gatePartitioners,
+            TaskInfo taskInfo) {
+        super(
+                checkpointedInputGate,
+                inputSerializer,
+                statusWatermarkValve,
+                inputIndex,
+                getRecordDeserializers(
+                        checkpointedInputGate,
+                        inputSerializer,
+                        ioManager,
+                        inflightDataRescalingDescriptor,
+                        gatePartitioners,
+                        taskInfo));
+        this.ioManager = ioManager;
+
+        LOG.info(
+                "Created demultiplexer for input {} from {}",
+                inputIndex,
+                inflightDataRescalingDescriptor);
+    }
+
+    private static <T>
+            Map<InputChannelInfo, DemultiplexingRecordDeserializer<T>> getRecordDeserializers(
+                    CheckpointedInputGate checkpointedInputGate,
+                    TypeSerializer<T> inputSerializer,
+                    IOManager ioManager,
+                    InflightDataRescalingDescriptor rescalingDescriptor,
+                    Function<Integer, StreamPartitioner<?>> gatePartitioners,
+                    TaskInfo taskInfo) {
+
+        RecordFilterFactory<T> recordFilterFactory =
+                new RecordFilterFactory<>(
+                        taskInfo.getIndexOfThisSubtask(),
+                        inputSerializer,
+                        taskInfo.getNumberOfParallelSubtasks(),
+                        gatePartitioners,
+                        taskInfo.getMaxNumberOfParallelSubtasks());
+        final DeserializerFactory deserializerFactory = new DeserializerFactory(ioManager);
+        Map<InputChannelInfo, DemultiplexingRecordDeserializer<T>> deserializers =
+                Maps.newHashMapWithExpectedSize(checkpointedInputGate.getChannelInfos().size());
+        for (InputChannelInfo channelInfo : checkpointedInputGate.getChannelInfos()) {
+            deserializers.put(
+                    channelInfo,
+                    DemultiplexingRecordDeserializer.create(
+                            channelInfo,
+                            rescalingDescriptor,
+                            deserializerFactory,
+                            recordFilterFactory));
+        }
+        return deserializers;
+    }
+
+    @Override
+    public StreamTaskInput<T> finishRecovery() throws IOException {
+        close();
```

Comment:
Do we need to make sure that all the buffers in deserializers are consumed?

Reply:
Yes good idea.